enterprise/cli/proxyserver.go

```diff
@@ -0,0 +1,361 @@
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/cli"
+	"github.com/coder/coder/cli/clibase"
+	"github.com/coder/coder/cli/cliui"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/coderd/httpmw"
+	"github.com/coder/coder/coderd/workspaceapps"
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/enterprise/wsproxy"
+)
+
+type closers []func()
+
+func (c closers) Close() {
+	for _, closeF := range c {
+		closeF()
+	}
+}
+
+func (c *closers) Add(f func()) {
+	*c = append(*c, f)
+}
+
+func (r *RootCmd) proxyServer() *clibase.Cmd {
+	var (
+		cfg = new(codersdk.DeploymentValues)
+		// Filter options for only relevant ones.
+		opts = cfg.Options().Filter(codersdk.IsWorkspaceProxies)
+
+		externalProxyOptionGroup = clibase.Group{
+			Name: "External Workspace Proxy",
+			YAML: "externalWorkspaceProxy",
+		}
+		proxySessionToken clibase.String
+		primaryAccessURL  clibase.URL
+		appSecuritYKey    clibase.String
+	)
+	opts.Add(
+		// Options only for external workspace proxies
+
+		clibase.Option{
+			Name:        "Proxy Session Token",
+			Description: "Authentication token for the workspace proxy to communicate with coderd.",
+			Flag:        "proxy-session-token",
+			Env:         "CODER_PROXY_SESSION_TOKEN",
+			YAML:        "proxySessionToken",
+			Default:     "",
+			Value:       &proxySessionToken,
+			Group:       &externalProxyOptionGroup,
+			Hidden:      false,
+		},
+
+		clibase.Option{
+			Name:        "Coderd (Primary) Access URL",
+			Description: "URL to communicate with coderd. This should match the access URL of the Coder deployment.",
+			Flag:        "primary-access-url",
+			Env:         "CODER_PRIMARY_ACCESS_URL",
+			YAML:        "primaryAccessURL",
+			Default:     "",
+			Value:       &primaryAccessURL,
+			Group:       &externalProxyOptionGroup,
+			Hidden:      false,
+		},
+
+		// TODO: This will eventually be pulled over an authenticated api endpoint.
+		clibase.Option{
+			Name:        "App Security Key",
+			Description: "App security key used for decrypting/verifying app tokens sent from coderd.",
+			Flag:        "app-security-key",
+			Env:         "CODER_APP_SECURITY_KEY",
+			YAML:        "appSecurityKey",
+			Default:     "",
+			Value:       &appSecuritYKey,
+			Group:       &externalProxyOptionGroup,
+			Hidden:      false,
+			Annotations: clibase.Annotations{}.Mark("secret", "true"),
+		},
+	)
+
+	client := new(codersdk.Client)
+	cmd := &clibase.Cmd{
+		Use:     "server",
+		Short:   "Start a workspace proxy server",
+		Options: opts,
+		Middleware: clibase.Chain(
+			cli.WriteConfigMW(cfg),
+			cli.PrintDeprecatedOptions(),
+			clibase.RequireNArgs(0),
+			// We need a client to connect with the primary coderd instance.
+			r.InitClient(client),
+		),
+		Handler: func(inv *clibase.Invocation) error {
+			if !(primaryAccessURL.Scheme == "http" || primaryAccessURL.Scheme == "https") {
```

Comment:
it'd be nice if deployment options can be validated in the option definition but alas

Reply:
Interesting idea... @ammario ? Is a "Validate" func field on an option reasonable?

Reply:
I like that idea